packages/grafana-runtime/src/context/GrafanaContext.ts

```diff
@@ -0,0 +1,23 @@
+import React, { useContext } from 'react';
+
+import { GrafanaConfig } from '@grafana/data';
+
+import { LocationService } from '../services/LocationService';
+import { BackendSrv } from '../services/backendSrv';
+
+export interface GrafanaContextType {
+  backend: BackendSrv;
+  location: LocationService;
+  config: GrafanaConfig;
+}
+
+export const GrafanaContext = React.createContext<GrafanaContextType | undefined>(undefined);
+
+export function useGrafana<TExtra extends object = {}>(): GrafanaContextType & TExtra {
+  const context = useContext(GrafanaContext);
+  if (!context) {
+    throw new Error('No GrafanaContext found');
+  }
+  // eslint-disable-next-line @typescript-eslint/consistent-type-assertions
+  return context as GrafanaContextType & TExtra;
+}
```

Comment:
I think this (and `useGrafanaInternal`) are simple enough that they can withstand repeating (if typed correctly!) and avoid the type assertion from the generic just to have a "private" API.

```jsx
interface PublicGrafanaContext {
  backend: BackendSrv;
  location: LocationService;
  config: GrafanaConfig;
}

function useGrafana(): PublicGrafanaContext {
  const context = useContext(GrafanaContext);
  if (!context) { throw ... }
  return { backend: context.backend, location: context.location, config: context.config }
}

function useGrafanaInternal(): GrafanaContextWithInternals {
  const context = useContext(GrafanaContext);
  if (!context) { throw ... }
  return context;
}
```

Reply:
yes, your right 👍